wurst/lib/ItemRecipe.wurst

```diff
@@ -0,0 +1,527 @@
+HashMap<int, ItemData> eventRecipeData = null
+trigger eventAssemblingTrigger = CreateTrigger()
+trigger eventAssembledTrigger = CreateTrigger()
+boolean eventHandled = false
+
+public constant int EVENT_ITEM_RECIPE_ASSEMBLING = 0
+public constant int EVENT_ITEM_RECIPE_ASSEMBLED = 1
+
+/** Returns triggering item recipe. */
+public function getEventItemRecipe() returns ItemRecipe
+    return eventRecipe
+
+/** Returns recipe triggering unit. */
+public function getEventItemRecipeUnit() returns unit
+    return eventUnit
+
+/** Returns event reward item. */
+public function getEventItemRecipeItem() returns item
+    return eventItem
+
+/** Returns unit item slots mapped with respective ingredients chosen for assembly. */
+public function getEventItemRecipeData() returns HashMap<int, ItemData>
+    return eventRecipeData
+
+/** Sets value indicating if event should be propagated. */
+public function setEventItemRecipeHandled(boolean handled)
+    eventHandled = handled
+
+/** Registers new event handler for specified item recipe event. */
+public function registerItemRecipeEvent(int whichEvent, code c)
+    if whichEvent == EVENT_ITEM_RECIPE_ASSEMBLING
+        eventAssemblingTrigger.addCondition(Condition(c))
+    else if whichEvent == EVENT_ITEM_RECIPE_ASSEMBLED
+        eventAssembledTrigger.addCondition(Condition(c))
+
+/** Returns trigger handle associated with specified recipe event. */
+public function getInventoryEventTrigger(int whichEvent) returns trigger
+    if whichEvent == EVENT_ITEM_RECIPE_ASSEMBLING
+        return eventAssemblingTrigger
+    else if whichEvent == EVENT_ITEM_RECIPE_ASSEMBLED
+        return eventAssembledTrigger
+    return null
+
+function fireEvent(trigger evt, ItemRecipe recipe, unit u, item itm, HashMap<int, ItemData> recipeData)
+    ItemRecipe prevRecipe = eventRecipe
+    unit prevUnit = eventUnit
+    item prevItem = eventItem
+    let prevRecipeData = eventRecipeData
+
+    eventRecipe = recipe
+    eventUnit = u
+    eventItem = itm
+    eventRecipeData = recipeData
+
+    evt.evaluate()
+
+    eventRecipe = prevRecipe
+    eventUnit = prevUnit
+    eventItem = prevItem
+    eventRecipeData = prevRecipeData
+
+/** Stores information about item required for recipe assembly. */
+public class ItemData
+    int itemTypeId
+    boolean remove
+    int charges
+    int index
+
+    construct(int itemTypeId, boolean remove, int charges, int index)
+        this.itemTypeId = itemTypeId
+        this.remove = remove
+        this.charges = charges
+        this.index = index
+    
+    construct(thistype base)
+        this.itemTypeId = base.itemTypeId
+        this.remove = base.remove
+        this.charges = base.charges
+        this.index = base.index
+
+/** Item recipe, collection of items that can be combined for powerful rewards. */
+public class ItemRecipe
+    private LinkedList<ItemData> ingredients
+    private int reward
+    int charges
+    boolean ordered
+    boolean permanent
```

Comment:
I think that we should collectively decide on a convention for booleans. I prefer `is` as a prefix, e.g. `isOrdered` and `isPermanent`, although `should` fits better in some cases.